File: lib/src/hlc.dart
import 'dart:math';

/// Hybrid Logical Clock implementation
/// 
/// Combines physical time (wall clock) with logical counter and node ID
/// to provide better ordering and causality detection than pure vector clocks.
/// 
/// Format: (physical_time_ms, logical_counter, node_id)
class HLC {
  /// Physical timestamp in milliseconds since Unix epoch
  final int physicalTime;
  
  /// Logical counter for events within the same millisecond
  final int logicalCounter;
  
  /// Node identifier - GUID v4 for users, hardcoded for services
  final String nodeId;

  const HLC({
    required this.physicalTime,
    required this.logicalCounter,
    required this.nodeId,
  });

  /// Create HLC with current wall clock time
  factory HLC.now(String nodeId) {
    return HLC(
      physicalTime: DateTime.now().millisecondsSinceEpoch,
      logicalCounter: 0,
      nodeId: nodeId,
    );
  }

  /// Create HLC from another HLC, incrementing logical counter
  HLC increment() {
    final currentTime = DateTime.now().millisecondsSinceEpoch;
    
    if (currentTime > physicalTime) {
      // Wall clock advanced, reset logical counter
      return HLC(
        physicalTime: currentTime,
        logicalCounter: 0,
        nodeId: nodeId,
      );
    } else {
      // Same millisecond, increment logical counter
      return HLC(
        physicalTime: physicalTime,
        logicalCounter: logicalCounter + 1,
        nodeId: nodeId,
      );
    }
  }

  /// Update HLC when receiving an event from another node
  HLC receiveEvent(HLC remoteHLC) {
    final currentTime = DateTime.now().millisecondsSinceEpoch;
    final newPhysicalTime = max(currentTime, max(physicalTime, remoteHLC.physicalTime));
    
    int newLogicalCounter;
    if (newPhysicalTime == remoteHLC.physicalTime && newPhysicalTime == physicalTime) {
      // All times equal, take max of logical counters and increment
      newLogicalCounter = max(logicalCounter, remoteHLC.logicalCounter) + 1;
    } else if (newPhysicalTime == remoteHLC.physicalTime) {
      // Remote time equals new time, increment remote logical counter
      newLogicalCounter = remoteHLC.logicalCounter + 1;
    } else if (newPhysicalTime == physicalTime) {
      // Local time equals new time, increment local logical counter
      newLogicalCounter = logicalCounter + 1;
    } else {
      // Physical time advanced beyond both, reset logical counter
      newLogicalCounter = 0;
    }

    return HLC(
      physicalTime: newPhysicalTime,
      logicalCounter: newLogicalCounter,
      nodeId: nodeId,
    );
  }

  /// Compare HLC for happens-before relationship
  /// Returns negative if this < other, positive if this > other, 0 if concurrent
  int compareTo(HLC other) {
    // First compare physical time
    final timeDiff = physicalTime.compareTo(other.physicalTime);
    if (timeDiff != 0) return timeDiff;
    
    // If physical time is equal, compare logical counter
    final logicalDiff = logicalCounter.compareTo(other.logicalCounter);
    if (logicalDiff != 0) return logicalDiff;
    
    // If both are equal, compare node IDs for deterministic ordering
    return nodeId.compareTo(other.nodeId);
  }

  /// Check if this HLC happens before another HLC
  bool happensBefore(HLC other) => compareTo(other) < 0;

  /// Check if this HLC happens after another HLC
  bool happensAfter(HLC other) => compareTo(other) > 0;

  /// Check if two HLCs are concurrent (neither happens before the other in a meaningful way)
  bool isConcurrentWith(HLC other) {
    return physicalTime == other.physicalTime && 
           logicalCounter == other.logicalCounter &&
           nodeId != other.nodeId;
  }

  /// Check if two HLCs represent the same logical time
  bool isEquivalentTo(HLC other) {
    return physicalTime == other.physicalTime &&
           logicalCounter == other.logicalCounter &&
           nodeId == other.nodeId;
  }

  @override
  bool operator ==(Object other) {
    return other is HLC && 
           other.physicalTime == physicalTime &&
           other.logicalCounter == logicalCounter &&
           other.nodeId == nodeId;
  }

  @override
  int get hashCode => Object.hash(physicalTime, logicalCounter, nodeId);

  @override
  String toString() => 'HLC($physicalTime:$logicalCounter@$nodeId)';

  /// Convert to JSON for serialization
  Map<String, dynamic> toJson() {
    return {
      'physicalTime': physicalTime,
      'logicalCounter': logicalCounter,
      'nodeId': nodeId,
    };
  }

  /// Create HLC from JSON
  static HLC fromJson(Map<String, dynamic> json) {
    return HLC(
      physicalTime: json['physicalTime'] as int,
      logicalCounter: json['logicalCounter'] as int,
      nodeId: json['nodeId'] as String,
    );
  }

  /// Create a copy of this HLC with different values
  HLC copyWith({
    int? physicalTime,
    int? logicalCounter,
    String? nodeId,
  }) {
    return HLC(
      physicalTime: physicalTime ?? this.physicalTime,
      logicalCounter: logicalCounter ?? this.logicalCounter,
      nodeId: nodeId ?? this.nodeId,
    );
  }
}

/// Generate a GUID v4 string for node identification
String generateGuidV4() {
  final random = Random.secure();
  final bytes = List<int>.generate(16, (_) => random.nextInt(256));
  
  // Set version (4) and variant bits according to RFC 4122
  bytes[6] = (bytes[6] & 0x0f) | 0x40; // Version 4
  bytes[8] = (bytes[8] & 0x3f) | 0x80; // Variant bits
  
  final hex = bytes.map((b) => b.toRadixString(16).padLeft(2, '0')).join();
  
  return '${hex.substring(0, 8)}-${hex.substring(8, 12)}-${hex.substring(12, 16)}-${hex.substring(16, 20)}-${hex.substring(20, 32)}';
}

File: csharp/YjsCrdtSharp/Core/HLC.cs
using System;
using System.Collections.Generic;

namespace YjsCrdtSharp.Core
{
    /// <summary>
    /// Hybrid Logical Clock implementation for causality tracking and ordering.
    /// Combines physical time (wall clock) with logical counter and node ID
    /// to provide better ordering and causality detection than pure vector clocks.
    /// </summary>
    public readonly struct HLC : IComparable<HLC>, IEquatable<HLC>
    {
        /// <summary>Physical timestamp in milliseconds since Unix epoch</summary>
        public long PhysicalTime { get; }
        
        /// <summary>Logical counter for events within the same millisecond</summary>
        public int LogicalCounter { get; }
        
        /// <summary>Node identifier - GUID v4 for users, hardcoded for services</summary>
        public string NodeId { get; }

        public HLC(long physicalTime, int logicalCounter, string nodeId)
        {
            PhysicalTime = physicalTime;
            LogicalCounter = logicalCounter;
            NodeId = nodeId ?? throw new ArgumentNullException(nameof(nodeId));
        }

        /// <summary>Create HLC with current wall clock time</summary>
        public static HLC Now(string nodeId)
        {
            return new HLC(
                physicalTime: DateTimeOffset.UtcNow.ToUnixTimeMilliseconds(),
                logicalCounter: 0,
                nodeId: nodeId
            );
        }

        /// <summary>Create HLC from another HLC, incrementing logical counter</summary>
        public HLC Increment()
        {
            var currentTime = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
            
            if (currentTime > PhysicalTime)
            {
                // Wall clock advanced, reset logical counter
                return new HLC(currentTime, 0, NodeId);
            }
            else
            {
                // Same millisecond, increment logical counter
                return new HLC(PhysicalTime, LogicalCounter + 1, NodeId);
            }
        }

        /// <summary>Update HLC based on receiving an event from another node</summary>
        public HLC ReceiveEvent(HLC other)
        {
            var currentTime = DateTimeOffset.UtcNow.ToUnixTimeMilliseconds();
            var maxTime = Math.Max(Math.Max(PhysicalTime, other.PhysicalTime), currentTime);
            
            int newLogicalCounter;
            if (maxTime == PhysicalTime && maxTime == other.PhysicalTime)
            {
                // Same millisecond as both events
                newLogicalCounter = Math.Max(LogicalCounter, other.LogicalCounter) + 1;
            }
            else if (maxTime == PhysicalTime)
            {
                // Our time is the maximum
                newLogicalCounter = LogicalCounter + 1;
            }
            else if (maxTime == other.PhysicalTime)
            {
                // Other time is the maximum
                newLogicalCounter = other.LogicalCounter + 1;
            }
            else
            {
                // Current time is the maximum
                newLogicalCounter = 0;
            }
            
            return new HLC(maxTime, newLogicalCounter, NodeId);
        }

        /// <summary>Check if this HLC happens before another</summary>
        public bool HappensBefore(HLC other)
        {
            if (PhysicalTime < other.PhysicalTime) return true;
            if (PhysicalTime > other.PhysicalTime) return false;
            if (LogicalCounter < other.LogicalCounter) return true;
            if (LogicalCounter > other.LogicalCounter) return false;
            return string.CompareOrdinal(NodeId, other.NodeId) < 0;
        }

        /// <summary>Check if this HLC happens after another</summary>
        public bool HappensAfter(HLC other) => other.HappensBefore(this);

        /// <summary>Compare two HLCs for ordering</summary>
        public int CompareTo(HLC other)
        {
            if (HappensBefore(other)) return -1;
            if (HappensAfter(other)) return 1;
            return 0;
        }

        /// <summary>Check equality with another HLC</summary>
        public bool Equals(HLC other)
        {
            return PhysicalTime == other.PhysicalTime &&
                   LogicalCounter == other.LogicalCounter &&
                   NodeId == other.NodeId;
        }

        public override bool Equals(object? obj) => obj is HLC other && Equals(other);

        public override int GetHashCode()
        {
            return HashCode.Combine(PhysicalTime, LogicalCounter, NodeId);
        }

        public static bool operator ==(HLC left, HLC right) => left.Equals(right);
        public static bool operator !=(HLC left, HLC right) => !(left == right);
        public static bool operator <(HLC left, HLC right) => left.HappensBefore(right);
        public static bool operator >(HLC left, HLC right) => left.HappensAfter(right);
        public static bool operator <=(HLC left, HLC right) => !left.HappensAfter(right);
        public static bool operator >=(HLC left, HLC right) => !left.HappensBefore(right);

        /// <summary>Convert to JSON for serialization</summary>
        public Dictionary<string, object> ToJson()
        {
            return new Dictionary<string, object>
            {
                ["physicalTime"] = PhysicalTime,
                ["logicalCounter"] = LogicalCounter,
                ["nodeId"] = NodeId
            };
        }

        /// <summary>Create HLC from JSON</summary>
        public static HLC FromJson(Dictionary<string, object> json)
        {
            return new HLC(
                physicalTime: Convert.ToInt64(json["physicalTime"]),
                logicalCounter: Convert.ToInt32(json["logicalCounter"]),
                nodeId: json["nodeId"].ToString()!
            );
        }

        public override string ToString()
        {
            return $"HLC({PhysicalTime}:{LogicalCounter}@{NodeId})";
        }
    }
}